torch/csrc/distributed/autograd/context/context.cpp

```diff
@@ -211,6 +213,15 @@ const c10::Dict<torch::Tensor, torch::Tensor> DistAutogradContext::
   return accumulatedGrads_;
 }
 
+void DistAutogradContext::runGradCallbaclForVariable(
+    const torch::autograd::Variable& variable,
+    GradCallback&& cb) {
+  std::lock_guard<std::mutex> guard(lock_);
+  auto it = accumulatedGrads_.find(variable);
+  TORCH_INTERNAL_ASSERT(it != accumulatedGrads_.end());
+  cb(it->value());
```

Comment:
We can release the lock before calling the callback?

Reply:
> We can release the lock before calling the callback?

We can't do that. `it` may be invalidated by another concurrent `insert()`.

Reply:
I can copy the value (a Tensor here) to avoid holding lock, but not sure whether it's possible that another thread may be accumulating to this grad. Anyway, the callsite from DDP reducer has a lightweight callback.